dbcon/mysql/ha_mcs_execplan.cpp

```diff
@@ -2906,15 +2906,36 @@ SimpleColumn* getSmallestColumn(boost::shared_ptr<CalpontSystemCatalog> csc,
 
             if (tan.alias == csep->derivedTbAlias())
             {
-                assert (!csep->returnedCols().empty());
-                ReturnedColumn* rc = dynamic_cast<ReturnedColumn*>(csep->returnedCols()[0].get());
+                const CalpontSelectExecutionPlan::ReturnedColumnList& cols = csep->returnedCols();
+
+                assert (!cols.empty());
+
+                CalpontSelectExecutionPlan::ReturnedColumnList::const_iterator iter;
+
+                ReturnedColumn* rc;
+
+                for (iter = cols.begin(); iter != cols.end(); iter++)
+                {
+                    if ((*iter)->refCount() != 0)
+                    {
+                        rc = dynamic_cast<ReturnedColumn*>(iter->get());
+                        break;
+                    }
+                }
+
+                if (iter == cols.end())
+                {
+                    rc = dynamic_cast<ReturnedColumn*>(cols[0].get());
```

Comment:
Plz add a comment explains that cols[0] is a dummy ConstantColumn made in derivedTableOptimization in the context of this block.

Reply:
Actually, cols[0] will not be a dummy ConstantColumn, since we increment its refCount here. An example query that triggers this codeblock:

```sql
SELECT "123" FROM (SELECT * FROM t1) q;
```

If table `t1` has two columns `a` and `b` in this order. And let's say `t1` has 10 records. Then rc here is a SimpleColumn referencing `a` for correct projection of 10 rows from the subquery.

Reply:
It is worth explaining for what reason do you blindly take cols[0].

Reply:
We take `cols[0]` here due to the optimization happening in derivedTableOptimization. All cols with refCount 0 from the end of the cols list are optimized out, until the first column with non-zero refCount is encountered. So here, if instead of `cols[0]`, we take `cols[1]` (based on some logic) and increment it's refCount, then `cols[0]` is not optimized out in derivedTableOptimization and is added as a ConstantColumn to the derived table's returned column list. This later causes an ineffective row group with row of the form (1, cols[1]_value1) to be created in ExeMgr.

Reply:
Very good. Could you put the same as the comment in the code?

Reply:
Sure, will add a comment here.